shot at bringing draft spec up to date with adrs (#44)
ga4gh/seqcol-spec: docs/specification.md

Table of contents:

[TOC]

## Specification version

This specification is in **DRAFT** form. This is **NOT YET AN APPROVED GA4GH specification**.

## Introduction

Reference sequences are fundamental to genomic analysis. To make analysis of reference sequences reproducible and efficient, we require tools that can identify, store, retrieve, and compare reference sequences. One such tool is the recent GA4GH standard [refget](http://samtools.github.io/hts-specs/refget.html), which provides a way to 1. compute deterministic sequence identifiers from sequences themselves, and 2) retrieve individual sequences using these unique identifiers derived. Refget thus provides a way to identify and retrieve sequences; however, many applications require working with *collections* of sequences. The *Sequence Collections* (seqcol) protocol is a pending GA4GH standard that works with collections of sequences. The seqcol protocol can be used to create unique identifiers derived from multiple sequences. Seqcol uses refget identifiers for individual sequences, adding new functionality to handle the complexity of collections, as well as attributes of sequences, such as names, lengths, or topologies. 

> The goal of the Seqcol project is **to standardize unique identifiers for collections of sequences**. It can be used to identify genomes, transcriptomes, or proteomes -- anything that can be represented as a collection of sequences. Seqcol uses a hash algorithm to generate a digest of the underlying collection of sequences. These unique identifiers are defined by an algorithm, rather than an accession authority, and are thus de-centralized and therefore usable for many purposes, including private or new sequence collections, cases without connection to a central database, or validation of sequence collection content and provenance.

Reviewer: I'm not sure I would emphasise uniqueness here: seqcol identifiers are not primarily unique. For me the main goal is to create identifiers derived from the collection itself.

```YAML
description: "A collection of biological sequences."
type: object
properties:
  lengths:
    type: array
    description: "Number of elements, such as nucleotides or amino acids, in each sequence."
    items:
      type: integer
  names:
    type: array
    description: "Human-readable identifiers of each sequence (chromosome names)."
    items:
      type: string
  sequences:
    type: array
    description: "Digests of sequences computed using the GA4GH digest algorithm (sha512t24u)."
    items:
      type: string
      description: "Actual sequence content"
required:
  - lengths
inherent:
  - lengths
  - names
  - sequences
```

We refer to this as the *seqcol object representation*. An example of a sequence collection organized into the seqcol object representation would be this:

```json
{
  "lengths": [
    8,
    4,
    4
  ],
  "names": [
    "chrX",
    "chr1",
    "chr2"
  ],
  "sequences": [
    "5f63cfaa3ef61f88c9635fb9d18ec945",
    "31fc6ca291a32fb9df82b85e5f077e31",
    "92c6a56c9e9459d8a42b96f7884710bc"

>   ],

Reviewer: ```suggestion
  ]
}
```

```

The object is a series of arrays with matching length (`3`), with the corresponding entries collated such that the first element of each array corresponds to the first element of each other array.

#### Seqcol algorithm:

Once the content of the sequence collection has been organized in the *seqcol object representation*, we can then run the digest algorithm to compute the unique identifier for this sequence collection

1. **Step 1**. Put all attributes of the sequence collection (*e.g.* names, lengths, sequence digests) in seqcol object representation form, as defined above.

> 2. **Step 2**. Apply [RFC-8785 JSON Canonicalization Scheme](https://www.rfc-editor.org/rfc/rfc8785) (JCS) to canonicalize the JSON of the seqcol object.

Reviewer: ```suggestion
2. **Step 2**. Apply [RFC-8785 JSON Canonicalization Scheme](https://www.rfc-editor.org/rfc/rfc8785) (JCS) to canonicalize the value associated with each attributes individually.
```

> 3. **Step 3**. Digest the canonical representation of each attribute using the seqcol digest algorithm (defined below). This should convert the value of each attribute in the seqcol into a digest string.

Reviewer: ```suggestion
3. **Step 3**. Digest the canonical representation of each attribute's value using the seqcol digest algorithm (defined below). This should convert the value of each attribute in the seqcol into a digest string.
```
